File: 1197-parsing-a-boolean-expression/parsing-a-boolean-expression.py
class Solution:
    def parseBoolExpr(self, expression: str) -> bool:
        i = 0

        def helper():
            nonlocal i
            c = expression[i]
            i += 1
            if c == 't': return True
            if c == 'f': return False
            i += 1
            if c == '!':
                res = not helper()
                i += 1
                return res
                

            elements = []
            while expression[i] != ')':
                if expression[i] != ',':
                    elements.append(helper())
                else:
                    i += 1
            i += 1
            if c == '&':
                return all(elements)
            if c == '|':
                return any(elements)
        return helper()

File: 1197-parsing-a-boolean-expression/parsing-a-boolean-expression.cpp
class Solution {
public:
    int i = 0;
    string str;
    bool helper(){
        char c= str[i];
        i += 1;
        if (c == 't') return true;
        if (c == 'f') return false;
        i += 1;
        if (c == '!'){
            bool res = ! helper();
            i += 1;
            return res;
        }
        vector<bool> container;
        while (str[i] != ')'){
            if (str[i] != ',') container.push_back(helper());
            else i += 1;
        }
        i += 1;
        bool result = container[0];
        if (c == '&'){
            for (int ii = 1; ii < container.size(); ii++) result &= container[ii];
        }
        if (c == '|'){
            for (int ii = 1; ii < container.size(); ii++) result |= container[ii];
        }

        return result;
    }
    bool parseBoolExpr(string expression) {
        str = expression;
        return helper();
    }
};